src/irmin/contents.ml

```diff
@@ -32,6 +111,139 @@ module Cstruct = struct
+    | `String s -> lexeme e (`String s)
+    | `Float f -> lexeme e (`Float f)
+    | `A a ->
+        lexeme e `As;
+        List.iter (encode_json e) a;
+        lexeme e `Ae;
+    | `O o ->
+        lexeme e `Os;
+        List.iter (fun (k, v) ->
+          lexeme e (`Name k);
+          encode_json e v
+        ) o;
+        lexeme e `Oe
+
+  let pp fmt x =
+    let buffer = Buffer.create 32 in
+    let encoder = Jsonm.encoder (`Buffer buffer) in
+    encode_json encoder (`O x);
+    ignore @@ Jsonm.encode encoder `End;
+    let s = Buffer.contents buffer in
+    Fmt.pf fmt "%s" s
+
+  let decode_json d =
+    let decode d = match Jsonm.decode d with
+      | `Lexeme l -> l
+      | `Error e -> failwith (Fmt.strf "%a" Jsonm.pp_error e)
+      | _ -> failwith "invalid JSON encoding"
+    in
+    let rec unwrap v d = match v with
+      | `Os -> obj [] d
+      | `As -> arr [] d
+      | `Null | `Bool _ | `String _ | `Float _ as v -> v
+      | _ -> failwith "invalid JSON value"
+    and arr vs d =
+      match decode d with
+      | `Ae -> `A (List.rev vs)
+      | v ->
+          let v = unwrap v d in
+          arr (v::vs) d
+    and obj ms d =
+      match decode d with
+      | `Oe -> `O (List.rev ms)
+      | `Name k ->
+          let v = unwrap (decode d) d in
+          obj ((k, v) :: ms) d
+      | _ -> failwith "invalid JSON object"
+    in
+    try
+      Ok (unwrap (decode d) d)
+    with
+      | Failure msg -> Error (`Msg msg)
+
+  let of_string s =
+    let decoder = Jsonm.decoder (`String s) in
+    match decode_json decoder with
+    | Ok (`O obj) -> Ok obj
+    | Ok _ -> Error (`Msg "Irmin JSON values must be objects")
+    | Error _ as err -> err
+end
+
+module Json_value = struct
+  type t = json
+  let t = json
+  let merge = Merge.(option merge_json)
+
+  let pp fmt x =
+    let buffer = Buffer.create 32 in
+    let encoder = Jsonm.encoder (`Buffer buffer) in
+    Json.encode_json encoder x;
+    ignore @@ Jsonm.encode encoder `End;
+    let s = Buffer.contents buffer in
+    Fmt.pf fmt "%s" s
+
+  let of_string s =
+    let decoder = Jsonm.decoder (`String s) in
+    match Json.decode_json decoder with
+    | Ok obj -> Ok obj
+    | Error _ as err -> err
+end
+
+module Json_tree(P: S.PATH)(M: S.METADATA) = struct
+  include Json_value
+
+  module type STORE = S.STORE with
+    type step = P.step
+    and type key = P.t
+    and type contents = json
+    and type metadata = M.t
+
+  let set_tree (type a) (type n) (module S: STORE with type t = a and type node = n)
+               (tree: S.tree) key (j : json) : S.tree Lwt.t =
+    match j with
+    | `O d ->
+        Lwt_list.fold_right_s (fun (k, v) tree ->
+          match S.Key.step_of_string k with
+          | Ok k ->
+            let k = S.Key.rcons key k in
+            S.Tree.add tree k v
+          | Error _ -> Lwt.return tree) d tree
+    | v -> S.Tree.add tree key v
+
+  let get_tree (type n) (module S: STORE with type node = n) (tree: S.tree) key =
```

Comment:
ha yes I see, having a first-class store is not very nice here. I am wondering if providing [Tree.concrete](https://mirage.github.io/irmin/irmin/Irmin/module-type-S/Tree/index.html#type-concrete) would just be enough for now.

Reply:
I was having some issues with [Tree.concrete]. It seems to only return the deepest node of the tree. Is this expected?

Reply:
That sounds like a bug!

Reply:
Okay, I will try to track that down as well.